core/capabilities/streams/mock_trigger.go

```diff
@@ -0,0 +1,472 @@
+	config     config
+}
+
+// Mock Trigger will send events to each subscriber every MaxFrequencyMs (configurable per subscriber).
+// Event generation happens whenever local unix time is a multiple of tickerResolutionMs. Therefore,
+// all subscribers' MaxFrequencyMs values need to be a multiple of tickerResolutionMs.
+func NewMockTriggerService(tickerResolutionMs int64, lggr logger.Logger) *MockTriggerService {
+	if tickerResolutionMs == 0 {
+		tickerResolutionMs = defaultTickerResolutionMs
+	}
+	//
+	f := 1
+	meta := datastreams.SignersMetadata{MinRequiredSignatures: 2*f + 1}
+	// gen private keys for MinRequiredSignatures
+	signers := []*ecdsa.PrivateKey{}
+	for i := 0; i < meta.MinRequiredSignatures; i++ {
+		// test keys: need to be the same across nodes
+		bytes := make([]byte, 32)
+		bytes[31] = uint8(i + 1)
+
+		privKey, err := crypto.ToECDSA(bytes)
+		if err != nil {
+			panic(err)
+		}
+		signers = append(signers, privKey)
+
+		signerAddr := crypto.PubkeyToAddress(privKey.PublicKey).Bytes()
+		meta.Signers = append(meta.Signers, signerAddr)
+	}
+	//
+	return &MockTriggerService{
+		Validator:          mercuryTriggerValidator,
+		CapabilityInfo:     capInfo,
+		tickerResolutionMs: tickerResolutionMs,
+		subscribers:        make(map[string]*subscriber),
+		latestReports:      make(map[datastreams.FeedID]datastreams.FeedReport),
+		stopCh:             make(services.StopChan),
+		lggr:               lggr.Named("MockTriggerService"),
+		meta:               meta,
+		signers:            signers}
+}
+
+func (o *MockTriggerService) ProcessReport(reports []datastreams.FeedReport) error {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	o.lggr.Debugw("ProcessReport", "nReports", len(reports))
+	for _, report := range reports {
+		feedID := datastreams.FeedID(report.FeedID)
+		o.latestReports[feedID] = report
+	}
+	return nil
+}
+
+func (o *MockTriggerService) RegisterTrigger(ctx context.Context, req capabilities.CapabilityRequest) (<-chan capabilities.CapabilityResponse, error) {
+	wid := req.Metadata.WorkflowID
+
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	config, err := o.ValidateConfig(req.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	inputs, err := o.ValidateInputs(req.Inputs)
+	if err != nil {
+		return nil, err
+	}
+
+	triggerID := o.getTriggerID(inputs.TriggerID, wid)
+	// If triggerId is already registered, return an error
+	if _, ok := o.subscribers[triggerID]; ok {
+		return nil, fmt.Errorf("triggerId %s already registered", triggerID)
+	}
+
+	if int64(config.MaxFrequencyMs)%o.tickerResolutionMs != 0 {
+		return nil, fmt.Errorf("MaxFrequencyMs must be a multiple of %d", o.tickerResolutionMs)
+	}
+
+	ch := make(chan capabilities.CapabilityResponse, defaultSendChannelBufferSize)
+	o.subscribers[triggerID] =
+		&subscriber{
+			ch:         ch,
+			workflowID: wid,
+			config:     *config,
+		}
+
+	// Only start the producer once a workflow is registered
+	o.producer = NewMockDataProducer(o, o.meta, o.signers, config.FeedIDs, o.lggr)
```

Comment:
Should we close the old producer before starting a new one, or just start it once on first registration?